docs/book/src/advanced/generics_and_trait_constraints.md

```diff
@@ -0,0 +1,125 @@
+# Generics and Trait Constraints
+
+## Generics as Constraints
+
+At a high level, Sway allows you to define constraints, or restrictions, that
+allow you to strike a balance between writing abstract and reusable code and
+enforcing compile-time checks to determine if the abstract code that you've
+written is correct.
+
+The "abstract and reusable" part largely comes from generic types and the
+"enforcing compile-time checks" part largely comes from trait constraints.
+Generic types can be used with functions, structs, and enums (as we have seen in
+this book), but they can also be used with traits.
+
+## Generic Traits
+
+Combining generic types with traits allows you to write abstract and reusable
+traits that can be implemented for any number of data types.
+
+For example, imagine that you want to write a trait for converting between
+different types. This would be similar to Rust's `Into` and `From` traits. In
+Sway your conversion trait would look something like:
+
+```sway
+trait Convert<T> {
+    fn from(t: T) -> Self;
+}
+```
+
+The trait `Convert` takes has a generic type `T`. `Convert` has one method
+`from`, which takes one parameter of type `T` and returns a `Self`. This means
+that when you implement `Convert` for a data type, `from` will return the type
+of that data type but will take as input the type that you define as `T`. Here
+is an example:
+
+```sway
```

Comment:
I would simplify the example to use 2 types instead of 3. I do not see the benefit of presenting another conversion which does not provide any different behaviour.